### PR TITLE
feat: add authz

### PR DESCRIPTION
**Pull request by Rory-Z**

### CODE REVIEW

File: apps/emqx_authorization/src/emqx_authorization.erl [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+
+-spec(register_metrics() -> ok).
+register_metrics() ->
+    lists:foreach(fun emqx_metrics:ensure/1, ?ACL_METRICS).
+
+init() ->
+    ok = register_metrics(),
+    % {ok, Conf} = hocon:load("etc/plugins/emqx_authorization.conf",#{format => richmap}),
+    RawConf = proplists:get_value(rules, application:get_all_env(?APP), []),
+    {ok, MapConf} = hocon:binary(jsx:encode(#{rules => RawConf}), #{format => richmap}),
```
Comment (zmstone) [inline apps/emqx_authorization/src/emqx_authorization.erl:38]: no need to encode and then load binary again if we call `hocon_schema:check_plain`

File: apps/emqx_authorization/src/emqx_authorization_schema.erl [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+-type action() :: pub | sub | pubsub.
+-type access() :: allow | deny.
+
+-reflect_type([ access/0
+              , action/0
+              ]).
+
+-export([structs/0, fields/1]).
+
+structs() -> ["rules"].
```
Comment (zmstone) [inline apps/emqx_authorization/src/emqx_authorization_schema.erl:14]:
```suggestion
structs() -> [""].
```

File: apps/emqx_authorization/src/emqx_authorization_schema.erl [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+fields("rules") ->
+    [ fun (type) -> {array, {union, ["base_rule", "connector_rule"]}};
+          (default) -> [];
+          (_) -> undefined
+      end
+    ];
```
Comment (zmstone) [inline apps/emqx_authorization/src/emqx_authorization_schema.erl:21]:
```suggestion
fields("") ->
    [ {rules, hoconsc:t({array,{union,["base_rule", "connector_rule"]}}, #{default => []}};
    ];
```

File: apps/emqx_authz/LICENSE [hunk 4]
```diff
@@ -0,0 +1,191 @@
+                                 Apache License
```
Comment (terry-xiaoyu) [inline apps/emqx_authz/LICENSE:1]: We may need to remove this LICENSE file

File: apps/emqx_authz/README.md [hunk 5]
```diff
@@ -0,0 +1,9 @@
+emqx_authz
```
Comment (terry-xiaoyu) [inline apps/emqx_authz/README.md:1]: The features provided by the app should be well documented in the readme.md

File: apps/emqx_authz/src/emqx_authz.app.src [hunk 6]
```diff
@@ -0,0 +1,17 @@
+{application, emqx_authz,
```
Comment (terry-xiaoyu) [inline apps/emqx_authz/src/emqx_authz.app.src:1]: It's hard to understand the word `authz`, if it is for ACL, then we could call the application as `emqx_acl`.

File: apps/emqx_authz/src/emqx_authz_api.erl [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    minirest:return(emqx_authz:update(NRules)).
+
+%%------------------------------------------------------------------------------
+%% Interval Funcs
+%%------------------------------------------------------------------------------
+
+get_rules(Params) ->
+    % #{<<"authz">> := #{<<"rules">> := Rules}} = hocon_schema:check_plain(emqx_authz_schema, #{<<"authz">> => Params}),
+    {ok, Conf} = hocon:binary(jsx:encode(#{<<"authz">> => Params}), #{format => richmap}),
+    CheckConf = hocon_schema:check(emqx_authz_schema, Conf),
```
Comment (terry-xiaoyu) [inline apps/emqx_authz/src/emqx_authz_api.erl:79]: now hocon_schem supports `check_plain` for configs that not richmap.

**Review by terry-xiaoyu — CHANGES_REQUESTED** [review]

File: apps/emqx_authz/src/emqx_authz.app.src (hunk 6 above)
Comment (Rory-Z) [reply apps/emqx_authz/src/emqx_authz.app.src:1]: @tigercl

File: apps/emqx_authz/src/emqx_authz_api.erl (hunk 7 above)
Comment (Rory-Z) [reply apps/emqx_authz/src/emqx_authz_api.erl:79]: But the `Params` is tuple list, so I'm not sure `check_plain` functions is useful

File: apps/emqx_authz/priv/emqx_authz.schema [hunk 8]
```diff
@@ -0,0 +1,8 @@
+%%-*- mode: erlang -*-
```
Comment (turtleDeng) [inline apps/emqx_authz/priv/emqx_authz.schema:1]: We need to remove this xxx.schema file

**Review by terry-xiaoyu — APPROVED** [review]

**Review by turtleDeng — APPROVED** [review]

**Review by terry-xiaoyu — APPROVED** [review]

**Pull request merged**